app/models/carto_uploader.rb

```diff
@@ -1,61 +1,141 @@
-  
   def upload(data)
-    allVals = []  #For Mass Insertion mode only.
-    keys = []     #Declaration for Mass Insertion mode.
+    #Determine what's already in the database
+    #--------------------------------
+    latest_classification_id = get_latest_classification_id()
+    expected_insertions = 0
+    successful_insertions = 0
+
+    #If we can't determine what's already in the database, DELETE AND START ANEW.
+    if latest_classification_id == 0
+      truncate_everything()
+    end
+    #--------------------------------
+
+    arr_vals = []
+    keys = []
     data.each do |classification|
+
+      #Only insert new Classifications
+      #--------------------------------
+      if Integer(classification["classification_id"]) <= latest_classification_id
+        next
+      end
+      #--------------------------------
+
+      #Take every key-val and package it into a single Classification to be Inserted
+      #--------------------------------
       keys = []  #Will be overwritten with the same data over and over again, but whatcha gonna do?
       vals = []
-      #Take every key-val 
-      #--------------------------------
       classification.each do |key, val|
         keys.push(key)
-        if (val.is_a?(NilClass))
+        if val.is_a?(NilClass)
           vals.push("''")
         else
           vals.push("'" + val.to_s.gsub(/'/, "''") + "'")
         end
       end
       keys = keys.join(",")
       vals = vals.join(",")
-      allVals.push("(#{vals})");  #For Mass Insertion mode only.
+      arr_vals.push("(#{vals})")
       #--------------------------------
 
-      #Option 1: Individual SQL Insertions
+      #If we have enough classifications queued up, do a batch Insert.
       #--------------------------------
-      if @mode != CARTODB_INSERT_MODE_MASS
-        sqlQuery = "INSERT INTO #{CARTODB_TABLE} (#{keys}) VALUES (#{vals}) "
-        uri = URI(CARTODB_URI_GET.gsub(/__SQLQUERY__/, URI.escape(sqlQuery)))
-        res = Net::HTTP.get(uri)
-        puts res
+      if arr_vals.length >= CLASSIFICATIONS_PER_BATCH
+        expected_insertions += arr_vals.length
+        successful_insertions += batch_insert(keys, arr_vals)
+        arr_vals = []
+        #Sanity check
+        puts "Inserted #{successful_insertions} Classifications out of #{expected_insertions}"
+        if expected_insertions != successful_insertions
+          raise StandardError, "Could not insert Classifications. Aborting attempt."
+        end
       end
       #--------------------------------
     end
 
-    #Option 2: Mass SQL Insertions
+    #Do a final batch Insert for any stragglers
     #--------------------------------
-    allVals = allVals.join(",")
-    if @mode == CARTODB_INSERT_MODE_MASS
-      sqlQuery = "INSERT INTO #{CARTODB_TABLE} (#{keys}) VALUES #{allVals} "
-      uri = URI(CARTODB_URI_POST)
-      res = Net::HTTP.post_form(uri, "q" => sqlQuery, "api_key" => CARTODB_APIKEY)
-      puts res.body
+    if !arr_vals.empty?
```

Comment:
While Hound is right, in this case I'd suggest `if arr_vals.present?` over `unless arr_vals.empty?`